### ISSUE TITLE
ybrid-uwsn-ml-framework

### ISSUE THREAD

**Issue by nahidhassanr26-create** [nahidhassanr26-create/hybrid_ml_framework.py#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from xgboost import XGBRegressor
from sklearn.metrics import r2_score
import os

# 1. Load the balanced hybrid underwater dataset
file_path = r"E:\comunication\data\balanced_hybrid_alpha_dataset_10000 - Copy.xlsx"
df = pd.read_excel(file_path)

# 2. Define the 11 Independent Input Features (X)
X_features = [
    'Distance_m', 'Turbidity_NTU', 'Depth_m', 'Temperature_C', 
    'Salinity_ppt', 'Alignment_Error_deg', 'Energy_Level_percent', 
    'Data_Size_KB', 'Popt_W', 'Pac_W', 'Alpha'
]
X = df[X_features]

# 3. Define the 12 Dependent Target Outputs (y)
y_targets = [
    'SNR_Optical_dB', 'SNR_Acoustic_dB', 'BER_Optical', 'BER_Acoustic',
    'Pr_opt_W', 'Pr_ac_W', 'Eopt_J', 'Eac_J', 
    'Lopt_s', 'Lac_s', 'Topt_bps', 'Tac_bps'
]
y = df[y_targets]

# 4. Perform an 80:20 Train-Test Stratification Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 5. Initialize the Three Machine Learning Architectures
rf_model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
lr_model = LinearRegression()
xgb_model = XGBRegressor(n_estimators=100, random_state=42, n_jobs=-1)

print("Training all three models simultaneously... Please wait a moment.")
# 6. Fit and Train the Architectures
rf_model.fit(X_train, y_train)
lr_model.fit(X_train, y_train)
xgb_model.fit(X_train, y_train)

# 7. Generate Multi-Output Predictions for Test Matrix
rf_pred = rf_model.predict(X_test)
lr_pred = lr_model.predict(X_test)
xgb_pred = xgb_model.predict(X_test)

rf_pred_df = pd.DataFrame(rf_pred, columns=y_targets)
lr_pred_df = pd.DataFrame(lr_pred, columns=y_targets)
xgb_pred_df = pd.DataFrame(xgb_pred, columns=y_targets)

# 📊 8. Evaluate and Build the Algorithmic Comparison Table (TABLE II)
comparison_results = []
for col in y_targets:
    rf_score = r2_score(y_test[col], rf_pred_df[col]) * 100
    lr_score = r2_score(y_test[col], lr_pred_df[col]) * 100
    xgb_score = r2_score(y_test[col], xgb_pred_df[col]) * 100
    
    comparison_results.append({
        'Target Parameter': col,
        'Linear Regression (R² %)': round(lr_score, 2),
        'XGBoost (R² %)': round(xgb_score, 2),
        'Random Forest (R² %)': round(rf_score, 2)
    })

summary_df = pd.DataFrame(comparison_results)
print("\n============= ML ALGORITHM COMPARISON RESULTS =============")
print(summary_df.to_string(index=False))
print("===========================================================\n")

# Save Table II to CSV format
summary_df.to_csv('ml_model_comparison_table.csv', index=False)
print("✓ TABLE II saved successfully as 'ml_model_comparison_table.csv'.")

# 📉 9. Generate and Save the 12 Individual Scatter Plots for Random Forest
print("\nGenerating and saving the 12 updated scatter plots...")
for col in y_targets:
    plt.figure(figsize=(6, 5))
    
    # Plot predicted teal scatter dots vs actual empirical baseline
    plt.scatter(y_test[col], rf_pred_df[col], color='darkcyan', alpha=0.5, edgecolors='none', label='Predicted Samples')
    
    # Draw the red dashed diagonal zero-error line
    plt.plot([y_test[col].min(), y_test[col].max()], [y_test[col].min(), y_test[col].max()], color='red', linestyle='--', lw=2, label='Perfect Fit')
    
    # Apply standard academic styling properties
    plt.xlabel(f'Actual {col}', fontsize=10)
    plt.ylabel(f'Predicted {col}', fontsize=10)
    plt.title(f'Updated Actual vs Predicted {col}', fontsize=11, fontweight='bold')
    plt.grid(True, linestyle=':', alpha=0.6)
    plt.legend(loc='upper left')
    plt.tight_layout()
    
    # Save each high-fidelity figure automatically to local folder
    filename = f"updated_plot_{col}.png"
    plt.savefig(filename, dpi=300)
    plt.close()
    print(f"✓ Saved figure: {filename}")

print("\n🎉 Process completed! Your algorithm comparison table and all 12 scatter plot images are fully ready.")
